app/http.php

```diff
@@ -0,0 +1,366 @@
+    Console::error('[Error] Line: ' . $error->getLine());
+}
+
+Console::success('Waiting for executors to start...');
+
+$startupDelay = \intval(App::getEnv('OPEN_RUNTIMES_PROXY_STARTUP_DELAY', '0'));
+if ($startupDelay > 0) {
+    \sleep((int) ($startupDelay / 1000));
+}
+
+\go(function () use ($register) {
+    if (App::getEnv('OPEN_RUNTIMES_PROXY_HEALTHCHECK', 'enabled') === 'enabled') {
+        fetchExecutorsState(true);
+    } else {
+        $executorStates = $register->get('executorStates');
+
+        // If no health check, mark all as online
+        $executors = \explode(',', (string) App::getEnv('OPEN_RUNTIMES_PROXY_EXECUTORS', ''));
+
+        foreach ($executors as $executor) {
+            $executorStates->set($executor, [
+                'status' => 'online',
+                'hostname' => $executor,
+                'state' =>  \json_encode([])
+            ]);
+        }
+    }
+});
+
+Console::log('State of executors at startup:');
+
+go(function () use ($register) {
+    $executorStates = $register->get('executorStates');
+
+    $executors = \explode(',', (string) App::getEnv('OPEN_RUNTIMES_PROXY_EXECUTORS', ''));
+
+    foreach ($executors as $executor) {
+        $state = $executorStates->exists($executor) ? $executorStates->get($executor) : null;
+
+        if ($state === null) {
+            Console::warning('Executor ' . $executor . ' has unknown state.');
+        } else {
+            Console::log('Executor ' . $executor . ' is ' . ($state['status'] ?? 'unknown') . '.');
+        }
+    }
+});
+
+Swoole\Event::wait();
+
+// Wildcard action
+App::error()
+    ->inject('request')
+    ->inject('response')
+    ->inject('roundRobinAtomic')
+    ->inject('executorStates')
+    ->action(function (Request $request, Response $response, Atomic $roundRobinAtomic, Table $executorStates) {
+        $secretKey = \explode(' ', $request->getHeader('authorization', ''))[1] ?? '';
+
+        if (empty($secretKey)) {
+            throw new Exception('Incorrect proxy key.', 401);
+        }
+        if ($secretKey !== App::getEnv('OPEN_RUNTIMES_PROXY_SECRET', '')) {
+            throw new Exception('Incorrect proxy key.', 401);
+        }
+
+        $roundRobinIndex = $roundRobinAtomic->get();
+
+        $algoType = App::getEnv('OPEN_RUNTIMES_PROXY_ALGORITHM', '');
+        $algo = match ($algoType) {
+            'round-robin' => new RoundRobin($roundRobinIndex - 1), // Atomic indexes from 1. Balancing library indexes from 0. That's why -1
+            'random' => new Random(),
+            default => new Random()
+        };
+
+        $balancing = new Balancing($algo);
+
+        $balancing->addFilter(fn ($option) => $option->getState('status', 'offline') === 'online');
+
+        foreach ($executorStates as $state) {
+            $balancing->addOption(new Option($state));
+        }
+
+        $body = \json_decode($request->getRawPayload() ? $request->getRawPayload() : '{}', true);
+        $runtimeId = $body['runtimeId'] ?? null;
```

Comment:
Can we get the runtime ID as a header and not in the payload? `X-OPR-Runtime`

What is the structure of the response we get from the executor? how much flexibility do we have?

Reply:
It is part of body because that is what executor expects.

Are you suggesting duplication (body for executor, header for proxy), or moving to header all-together even for executor?

Reply:
Not sure I fully understand.

Reply:
I made a comment out of it. This is TODO anyway.
This will become relevant when we will want to prefer warm runtimes (those with no cold start). That we don't plan to do as part of the initial implementation.